text/0003-font-types.md

```diff
@@ -0,0 +1,408 @@
+general case. In most cases, the user will be using one of these two sets of
+types, and in the advanced cases (e.g. server-side subsetting, instancing
+variable fonts) it still shouldn't be that complex.
+
+If it turns out that this approach does not meet our performance
+needs in certain special circumstances, we can consider having additional API
+for these on a per-case basis.
+
+## Extensibility
+
+A final element of the design I would like to highlight is extensibility. The
+basic idea here is that tables would not need to all be defined in a single
+canonical place. If a user wanted to add support for a new, non-standard table,
+all of the macros required to describe it would be available publicly, and they
+would only need to write out and annotate the structs.
+
+# Prior art
+
+There are a number of existing projects that I think are relevant to this work,
+and which encourage me as to its feasibility. The first of these is
+[fonttools-rs][], and its components
+otspec and otspec-macros. This is a set of crates by Simon Cozens that are
+focused on the compilation case (data is stored in native rust types and
+converted to raw bytes during serialization and deserialization) but it uses a
+system of macros to generate that IO code, to good effect. The second of these
+is Chad Brokow’s [pinot][], which is an existing
+zero-copy OpenType parser intended to be used for shaping.
+
+Between these two projects, the majority of the goals i have for this project
+have already been demonstrated to be feasible; the challenge here will be
+largely to figure out a way to combine them in a sensible and ergonomic way.
+
+Outside of Rust, I think it will also be important to spend some quality time
+with HarfBuzz, to make sure there are not requirements of that use-case that I
+am overlooking.
+
+
+# Schedule and time allocation
+
+The initial goal for this work is to determine whether it is feasible at all.
+With that in mind, I think we can break the project up into three phases, where
+we can abandon the work if any particular phase is a failure:
+
+
+1. macros, zero-copy types: demonstrate that we can use macros to generate code
+   for interpreting and accessing values from a small handful of tables,
+   including some tricky tables like GPOS/GSUB. 4-6 weeks.
+2. mutability: decide on a strategy for mutability, and implement it for that
+   same handful of tables. 6-10 weeks. It is possible that as we work on
+   mutability we will decide that we need to revisit how we structure our
+   code-gen code: for instance if we need to generate multiple types for each
+   table, a procedural macro may not be the best choice.
+3. Implement all core tables: ideally this is mostly just a matter of typing,
+   but it might still be quite a lot of typing. I’ll feel more comfortable
+   giving this a time estimate after I’ve finished the previous items, but if I
+   had to guess I would say ~12 weeks.
+
```

Comment:
I'd be very confident adding a step 4 which is to build some partial, but realistic, implementations of things, ideally sufficient to prove 1) we can do Real Work and 2) to gain an early sense of performance